zcommon/src/org/zkoss/video/AVideo.java

```diff
@@ -0,0 +1,259 @@
+import org.zkoss.lang.SystemException;
+import org.zkoss.util.media.ContentTypes;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.net.URL;
+
+/**
+ * Represents a video.
+ * <p>
+ * it is used to retrieve and store the opaque data
+ * as polymorphic thru the {@link org.zkoss.util.media.Media} interface.
+ * <p>
+ * <p>AVideo is serializable, but, if you are using InputStream,
+ * you have to extend this class, and provide the implementation to
+ * serialize and deserialize {@link #_isdata}. (Since 8.5.2)
+ */
+public class AVideo implements Video, java.io.Serializable {
+	/**
+	 * Used if you want to implement a media whose input stream is created
+	 * dynamically each time {@link #getStreamData} is called.
+	 *
+	 * @see #AVideo(String, , InputStream)
+	 */
+	protected static final InputStream DYNAMIC_STREAM = new NullInputStream();
+	/**
+	 * The raw data in byte array.
+	 * Exactly one of {@link #_data} and {@link #_isdata} is not null.
+	 */
+	private final byte[] _data;
+	/**
+	 * The raw data in stream (or {@link #DYNAMIC_STREAM}).
+	 * Exactly one of {@link #_data} and {@link #_isdata} is not null.
+	 */
+	protected final transient InputStream _isdata;
+	/**
+	 * The URL of the data.
+	 */
+	private final URL _url;
+	/**
+	 * The file of the data.
+	 */
+	private final File _file;
+	/**
+	 * The format name, e.g., "mp4", "webm" and "ogg".
+	 */
+	private String _format;
+	/**
+	 * The content type.
+	 */
+	private String _ctype;
+	/**
+	 * The name (usually filename).
+	 */
+	private final String _name;
+	
+	public AVideo(String name, byte[] data) throws IOException {
+		if (data == null)
+			throw new IllegalArgumentException("null data");
+		_name = name;
+		_data = data;
+		_isdata = null;
+		_url = null;
+		_file = null;
+	}
+	
+	/**
+	 * Creates an instance of a video with an input stream.
+	 * If the stream shall be created each time {@link #getStreamData} is called,
+	 * you can pass {@link #DYNAMIC_STREAM} to the data argument, and then
+	 * override {@link #getStreamData}.
+	 * <p>
+	 * <p>Note: the caller of {@link #getStreamData} has to close
+	 * the returned input stream.
+	 */
+	public AVideo(String name, InputStream isdata) throws IOException {
+		if (isdata == null)
+			throw new IllegalArgumentException("null stream");
+		_name = name;
+		_isdata = isdata;
+		_data = null;
+		_url = null;
+		_file = null;
+	}
+	
+	/**
+	 * Constructs a video with an URL.
+	 */
+	public AVideo(URL url) {
+		if (url == null)
+			throw new IllegalArgumentException("null url");
+		_name = getName(url);
+		_url = url;
+		_isdata = DYNAMIC_STREAM;
+		_data = null;
+		_file = null;
+	}
+	
+	/**
+	 * Constructs a video with a file.
+	 */
+	public AVideo(File file) {
+		if (file == null)
+			throw new IllegalArgumentException("null url");
```

Comment:
It is file, not url.
how about "file cannot be null"? that would be a completed sentence

Reply:
ok